File: 백준/Silver/1966. 프린터 큐/프린터 큐.py
from collections import deque


T = int(input())
for _ in range(T):
    N, M = map(int, input().split())
    arr = list(map(int, input().split()))
    q = deque(arr)

    # 큐의 최초 최대값 초기화
    max_q = max(q)
    result = 0
    # 우선 큐가 빌 때 까지는 진행
    while q:
        # 만약 q[0] 이 최댓값이 아니라면 뒤로 넘기고 넘어감
        if q[0] != max_q:
            q.rotate(-1)
            # 타겟하고 있는 문서를 추척
            if M != 0:
                M -= 1
            else:
                M = N - 1
            continue
        # 만약 최대값이라면
        # 최대값이면서 타겟하고 있는 문서가 아닌 경우
        docs = q.popleft()
        if M > 0:
            N -= 1
            M -= 1
            # 최댓값 갱신
            max_q = max(q)
            result += 1
        # 최댓값이면서 타겟하고 있는 문서인 경우
        elif M == 0:
            result += 1
            break

    print(result)

File: 백준/Silver/1966. 프린터 큐/프린터 큐.java
import java.io.*;
import java.lang.reflect.Array;
import java.util.*;

class Main {
    public static void main(String[] args) throws IOException {
//        System.setIn(new FileInputStream("input.txt"));
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();

        int T = Integer.parseInt(br.readLine());

        for (int i=0; i<T; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            // 문서의 개수
            int N = Integer.parseInt(st.nextToken());
            // 타겟 문서
            int M = Integer.parseInt(st.nextToken());

            StringTokenizer st2 = new StringTokenizer(br.readLine());

            Queue<Integer> q = new LinkedList<>();
            int maxNum = 0;
            // N만큼 순회하며 큐에 값 채우기
            for (int j=0; j<N; j++) {
                int nowNum = Integer.parseInt(st2.nextToken());
                q.add(nowNum);
                // 최댓값 갱신
                if (nowNum > maxNum) {
                    maxNum = nowNum;
                }
            } // 큐 채우기
            // 큐를 다 채웠다면 하나씩 꺼내서 maxNum인지 확인하기
            int result = 0;
            while (true) {
                // 큐에서 값을 하나 꺼냄
                int nowNum = q.poll();
                // 가장 큰 우선순위가 아닌경우
                if (nowNum != maxNum) {
                    q.add(nowNum);
                    if (M == 0) {
                        M = q.size() - 1;
                        continue;
                    } 
                    M -= 1;
                    // 가장 큰 우선순위를 꺼낸 경우
                } else if (nowNum == maxNum) {
                    // 타겟값과 일치하는지 확인
                    if (M == 0) {
                        result += 1;
                        break;
                } else {
                        M -= 1;
                        maxNum = Collections.max(q);
                        result += 1;
                    }

            }
            }
            sb.append(result).append("\n");
        } // 문제 for
        System.out.println(sb);
    }
}